### PR TITLE
Further optimize build status generation

### PR DESCRIPTION
**Pull request by ErikSchierboom**
- Optimize counting of mentor discussions
- Remove duplicate query from build status updating
- Greatly optimize user queries

### CODE REVIEW

File: app/commands/track/update_build_status.rb [hunk 1]
```diff
@@ -91,7 +91,7 @@ def num_submissions = track.submissions.count
 
   def mentor_discussions
     {
-      num_discussions: Mentor::Discussion.joins(:request).where(request: { track: }).count
+      num_discussions: track.mentor_requests.fulfilled.count
```
Comment (ErikSchierboom) [inline app/commands/track/update_build_status.rb:94]: Looking at the code, `fulfilled` seems to imply that a discussion was created for the request, If true, this query does not need a join and is quicker.

File: app/commands/track/update_build_status.rb [hunk 2]
```diff
@@ -141,19 +141,14 @@ def test_runner_version_target
 
   def representer
     {
-      num_runs: representer_num_runs,
+      num_runs: representer_num_submissions,
```
Comment (ErikSchierboom) [inline app/commands/track/update_build_status.rb:144]: The submissions representations were counted twice. This fixes that

File: app/commands/track/update_build_status.rb [hunk 3]
```diff
@@ -223,13 +218,15 @@ def syllabus_health
   end
 
   def syllabus_volunteers
-    authors = User.where(id: Concept::Authorship.where(concept: taught_concepts).select(:user_id)).
-      or(User.where(id: Exercise::Authorship.where(exercise: active_concept_exercises).select(:user_id)))
+    concept_author_ids = Concept::Authorship.where(concept: taught_concepts).distinct.pluck(:user_id)
+    exercise_author_ids = Exercise::Authorship.where(exercise: active_concept_exercises).distinct.pluck(:user_id)
+    author_ids = (concept_author_ids | exercise_author_ids).uniq
```
Comment (ErikSchierboom) [inline app/commands/track/update_build_status.rb:223]: The new pattern to serialize volunteers is to fetch the unique ids first via individual queries, and then to randomize and fetch the actual users using those ids.

File: app/commands/track/update_build_status.rb [hunk 4]
```diff
@@ -391,15 +388,33 @@ def serialize_prob_specs_exercise(exercise)
...
+    serialize_volunteers(author_ids, contributor_ids)
   end
 
-  def serialize_volunteers(authors, contributors, user_id_column: :id)
+  def serialize_volunteers(author_ids, contributor_ids)
+    random_author_ids = author_ids.shuffle.take(NUM_VOLUNTEERS)
+    random_authors = User.where(id: random_author_ids).to_a
+
+    random_contributor_ids = (contributor_ids - author_ids).shuffle.take(NUM_VOLUNTEERS - author_ids.size)
+    random_contributors = User.where(id: random_contributor_ids).to_a
```
Comment (ErikSchierboom) [inline app/commands/track/update_build_status.rb:411]: If there are three authors, `random_contributor_ids` will be an empty array causing `User.where(id: random_contributor_ids)` to be a noop.

File: app/commands/track/update_build_status.rb [hunk 5]
```diff
@@ -391,15 +388,33 @@ def serialize_prob_specs_exercise(exercise)
...
+
+    num_users = (author_ids | contributor_ids).uniq.size
```
Comment (ErikSchierboom) [inline app/commands/track/update_build_status.rb:413]: No additional count query needed

File: app/commands/track/update_build_status.rb (hunk 1 above)
Comment (iHiD) [reply app/commands/track/update_build_status.rb:94]: Nice.

File: app/commands/track/update_build_status.rb (hunk 3 above)
Comment (iHiD) [reply app/commands/track/update_build_status.rb:223]: Good.

File: app/commands/track/update_build_status.rb (hunk 5 above)
Comment (iHiD) [reply app/commands/track/update_build_status.rb:413]: Love a bit of `|`

Comment (iHiD) [reply app/commands/track/update_build_status.rb:413]: Another way of doing this that I do elsewhere is to create a Set

**Review by iHiD — APPROVED** [review]
Scanned it. Looks solid. LGTM!

File: app/commands/track/update_build_status.rb (hunk 5 above)
Comment (ErikSchierboom) [reply app/commands/track/update_build_status.rb:413]: Yes that would be an option too. I didn't know what was idiomatic 🤷

**Pull request merged**